truffle/src/main/java/org/jruby/truffle/core/regexp/RegexpNodes.java

```diff
@@ -426,7 +430,25 @@ public Object matchGeneric(VirtualFrame frame, DynamicObject regexp, DynamicObje
                 toStrNode = insert(ToStrNodeGen.create(getContext(), getSourceSection(), null));
             }
 
-            return match(regexp, toStrNode.executeToStr(frame, other));
+            return matchWithStringCopy(regexp, toStrNode.executeToStr(frame, other));
+        }
+
+        // Creating a MatchData will store a copy of the source string. It's tempting to use a rope here, but a bit
+        // inconvenient because we can't work with ropes directly in Ruby (useful for implementing some MatchData
+        // methods). Likewise, we need to taint objects based on the source string's taint state. We mustn't allow the
+        // source string's contents to change, however, so we must ensure that we have a private copy of that string.
+        private Object matchWithStringCopy(DynamicObject regexp, DynamicObject string) {
+            final Matcher matcher = createMatcher(getContext(), regexp, string);
+            final int range = StringOperations.rope(string).byteLength();
+
+            final DynamicObject matchData = matchCommon(getContext(), this, makeSubstringNode, regexp, string, true, matcher, 0, range);
+            setLastMatchNode.executeSetLastMatch(matchData);
+
+            if (matchData != nil()) {
+                return matcher.getBegin();
+            }
```

Comment:
That was what I thought originally and why I introduced the bug. But the following fragment shows how that can fall apart:

```
x = "abc"
x =~ /(.*)/
x.upcase!
```

Without making a defensive copy, `$~.string` will be `"ABC"`, rather than `"abc"`, which is what the match was really made against.